File: Etapa3-Construcao/Stack2/hotel-booking-java/src/main/java/com/hotel/booking/valueobjects/Endereco.java
package com.hotel.booking.valueobjects;

import jakarta.persistence.Embeddable;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Embeddable
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Endereco {
    private String logradouro;
    private String numero;
    private String complemento;
    private String bairro;
    private String cidade;
    private String estado;
    private String cep;
    private String pais;

    public boolean ehValido() {
        return logradouro != null && !logradouro.trim().isEmpty() &&
               numero != null && !numero.trim().isEmpty() &&
               bairro != null && !bairro.trim().isEmpty() &&
               cidade != null && !cidade.trim().isEmpty() &&
               estado != null && !estado.trim().isEmpty() &&
               cep != null && validarCEP(cep);
    }

    private boolean validarCEP(String cep) {
        return cep.matches("^\\d{5}-\\d{3}$");
    }

    public String getEnderecoCompleto() {
        StringBuilder endereco = new StringBuilder();
        endereco.append(logradouro).append(", ").append(numero);
        
        if (complemento != null && !complemento.trim().isEmpty()) {
            endereco.append(" - ").append(complemento);
        }
        
        endereco.append(", ").append(bairro);
        endereco.append(", ").append(cidade).append(" - ").append(estado);
        endereco.append(", ").append(cep);
        
        return endereco.toString();
    }
}

File: Etapa3-Construcao/Stack1/hotel-booking-system/src/domain/value-objects/Endereco.ts
// Value Object: Endereço
export class Endereco {
  constructor(
    public readonly logradouro: string,
    public readonly numero: string,
    public readonly complemento: string = '',
    public readonly bairro: string,
    public readonly cidade: string,
    public readonly estado: string,
    public readonly cep: string,
    public readonly pais: string = 'Brasil'
  ) {}

  // Validação básica do CEP
  private validarCEP(cep: string): boolean {
    const cepRegex = /^\d{5}-\d{3}$/;
    return cepRegex.test(cep);
  }

  // Formatação do endereço completo
  get enderecoCompleto(): string {
    let endereco = `${this.logradouro}, ${this.numero}`;
    
    if (this.complemento) {
      endereco += ` - ${this.complemento}`;
    }
    
    endereco += `, ${this.bairro}`;
    endereco += `, ${this.cidade} - ${this.estado}`;
    endereco += `, ${this.cep}`;
    
    return endereco;
  }

  // Verifica se o endereço é válido
  ehValido(): boolean {
    return (
      this.logradouro.trim().length > 0 &&
      this.numero.trim().length > 0 &&
      this.bairro.trim().length > 0 &&
      this.cidade.trim().length > 0 &&
      this.estado.trim().length > 0 &&
      this.validarCEP(this.cep)
    );
  }

  // Cria uma cópia do endereço com novos valores
  atualizar(dados: Partial<Endereco>): Endereco {
    return new Endereco(
      dados.logradouro ?? this.logradouro,
      dados.numero ?? this.numero,
      dados.complemento ?? this.complemento,
      dados.bairro ?? this.bairro,
      dados.cidade ?? this.cidade,
      dados.estado ?? this.estado,
      dados.cep ?? this.cep,
      dados.pais ?? this.pais
    );
  }

  // Converte para objeto JSON
  toJSON() {
    return {
      logradouro: this.logradouro,
      numero: this.numero,
      complemento: this.complemento,
      bairro: this.bairro,
      cidade: this.cidade,
      estado: this.estado,
      cep: this.cep,
      pais: this.pais
    };
  }

  // Cria instância a partir de objeto JSON
  static fromJSON(dados: any): Endereco {
    return new Endereco(
      dados.logradouro,
      dados.numero,
      dados.complemento || '',
      dados.bairro,
      dados.cidade,
      dados.estado,
      dados.cep,
      dados.pais || 'Brasil'
    );
  }
}
